aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -1162,6 +1162,81 @@ Tensor take_backward(const Tensor& grad, const Tensor& input, const Tensor& inde
   return at::zeros_like(input).put_(index, grad, true);
 }
 
+namespace {
+
+inline std::tuple<Tensor, Tensor, int64_t> _take_along_dim_helper(
+    const Tensor& self,
+    const Tensor& indices,
+    int64_t dim) {
+  TORCH_CHECK(
+      self.dim() == indices.dim(),
+      "torch.take_along_dim(): input and indices should have the same number of dimensions, ",
+      "but got ", self.dim(), " dimensions for input, and ", indices.dim(), " dimensions for indices")
+  TORCH_CHECK(
+      indices.scalar_type() == ScalarType::Long,
+      "torch.take_along_dim(): dtype of indices should be Long but got ", indices.scalar_type())
+
+  dim = at::maybe_wrap_dim(dim, self.dim());
+
+  DimVector self_sizes{self.sizes()};
+  // update number of elements at dim as per indices
+  self_sizes[dim] = indices.size(dim);
+  auto broadcast_shape = infer_size(self_sizes, indices.sizes());
+  auto indices_broadcasted = at::broadcast_to(indices, broadcast_shape);
+
+  DimVector indices_sizes{indices.sizes()};
+  // update number of elements at dim as per self
+  indices_sizes[dim] = self.size(dim);
+  broadcast_shape = infer_size(indices_sizes, self.sizes());
+  auto self_broadcasted = at::broadcast_to(self, broadcast_shape);
+
+  return std::make_tuple(self_broadcasted, indices_broadcasted, dim);
+}
+
+static inline void checkDevice(CheckedFrom c, const Tensor& t, Device device) {
```

Comment:
Should this belong here?
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorUtils.cpp

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorUtils.h

The file has `checkDeviceType` which only checks for device type but does not consider the index of the device.
However functions are marked with TORCH_API (not sure what this exactly implies) so not sure if we should add this :thinking:

Reply:
Reference to discuss: https://discuss.pytorch.org/t/what-is-torch-api/67398

Reply:
Maybe. Let's think about improving our guidance for how ops should perform error checking separately, and then we can identify the best pattern. These device checks are typically simple to write, so I'm not sure a helper would add much.